[WIP] Improve use of git revision

eksperimental opened the pull request:
This PR does a few things all related to the git version that is shown in the elixir version

- "mix --version" shows the hash id, same as elixir and iex
- A new a function added (System.build_info/1) retrieves the string to display in "elixir --version" and friends.
- System.read_stripped/1 fixed to properly deal with \r, \n and \t (not only with white spaces). This solves the problem that VERSION cannot end in a new line
- System.get_revision/0 returns the long hash now (instead of the short one)
I think it's better to return the original one, and then when we display it shortening it, as it may be needed, for example linking to the github commit_id which uses the long one.
this could be considered a backward incompatible change.
- System.get_revision/0 also returns the git tag.
- Make file, functions created to retrieve the git tag and revision
- Makefile: in the current version `git` is a requirement if we want to build docs (when it should not), and if git is not installed, Makefile will complain. With these changes, git will fail silently if not installed.

whatyouhide commented on lib/mix/lib/mix/cli.ex:
```diff
@@ -114,7 +114,8 @@ defmodule Mix.CLI do
 
   defp display_version() do
-    IO.puts "Mix #{System.version}"
+    #IO.puts :erlang.system_info(:system_version)
```
:scissors: ?

eksperimental replied:
oh.. yeah. thanks for reminding me that one.

```sh
$ elixir --version
Erlang/OTP 18 [erts-7.0] [source] [smp:2:2] [async-threads:10] [hipe] [kernel-poll:false]

Elixir 1.3.0-dev (b587871)
$ iex --version
Erlang/OTP 18 [erts-7.0] [source] [smp:2:2] [async-threads:10] [hipe] [kernel-poll:false]

IEx 1.3.0-dev
$ mix --version
Mix 1.3.0-dev
```

Should we make it consistent with the other commands?

eksperimental replied:
any thoughts on this one? @lexmag @josevalim @whatyouhide

josevalim replied:
It is ok to make it consistent.

lexmag replied:
Consistency :yellow_heart:.

josevalim commented:
Thank you @eksperimental. There are too many changes in here, so it makes it hard to track what is really happening. I will add some comments with questions.

josevalim commented on Makefile:
```diff
@@ -11,6 +11,8 @@ INSTALL = install
 INSTALL_DIR = $(INSTALL) -m755 -d
 INSTALL_DATA = $(INSTALL) -m644
 INSTALL_PROGRAM = $(INSTALL) -m755
+GIT_REVISION = $(shell (git rev-parse HEAD 2> /dev/null) || (cat .git/HEAD 2> /dev/null | head -c 40))
```
Why do we need to do all this dance given we didn't need it before?

eksperimental replied:
the commands are different, before it was only `git rev-parse HEAD`, but I have mimicked how we do it in the System.get_revision/0 function.

One important point I forgot to mention in the PR is that in the current MakeFile, `git` is a requirement if we want to build docs (when it should not), and if git is not installed, Makefile will complain.
```sh
/bin/sh: 1: git: not found
/bin/sh: 1: git: not found
```

With these changes, it will just fail silently.

josevalim replied:
Perfect, thank you.

josevalim replied:
On the other hand, I think it is fine to have git as a dependency for some make file commands, no? Otherwise, we will generate the wrong docs anyway.

lexmag replied:
I'm for having git as a dependency.

eksperimental replied:
@josevalim: you are correct about generating the wrong docs, since it will be linking to master.

Actually, this PR came initially as a one to avoid using git as a dependency, and the way I found to deal with this was to generate to files `.git_revision` and `.git_tag` that include the information there, provided git was not installed. What made me give up on this idea was that github releases doesn't allow (AFAIK) a way to execute a command when packing the release, and to generate `.git_revision` on the fly.
It's sort of pointless to make git a dependency when all we need is the revision id, but the only way solve this is a upload the release file ourselves.

One scenario that we haven't considered, is that even if git is present will not work is when someone dowloads a zip file with the source code, and tries to build the docs... there is no git information anywhere... so won't be able to build it. Therefore I think it makes sense in some way to build the release ourselves and upload, and not have git as a dependency.
what do you guys think?

josevalim commented on lib/elixir/lib/system.ex:
```diff
@@ -5,40 +5,71 @@ defmodule System do
-          strip_re(data, "\n")
+          :os.cmd('git rev-parse HEAD') |> strip
+        else
+          read_stripped(@git_head_file)
```
why?

josevalim commented on lib/elixir/lib/system.ex:
```diff
@@ -5,40 +5,71 @@ defmodule System do
+  end
+
+  @spec get_tag :: String.t
+  defmacrop get_tag do
```
I am :-1:. Why include it? What is the use case? Often the tag will be the version unless the version is dev. I don't think this is adding anything beyond code complexity.

eksperimental replied:
Use case is up to the user. Let's say somebody wants to link to github to the tag, but if no tag given, then link to the revision id. same as we do in the makefile when we build the docs and link to the source files.

josevalim replied:
:-1:. If someone has a use case, then please come and ask. I am really not a fan of guessing possible use cases upfront.

josevalim replied:
Also, if there is a tag, it is very very very likely that it will be the Elixir version.

lexmag replied:
Best code is the code you avoided writing in the first place. :bowtie:

eksperimental replied:
you guys are definitely right.

josevalim commented on lib/elixir/lib/system.ex:
```diff
@@ -73,11 +104,30 @@ defmodule System do
   @spec build_info() :: map
   def build_info do
-    %{version: version, date: get_date, revision: get_revision}
+    %{  date:     get_date,
```
Please fix the indentation. Also, we should leave it as `version` instead of `get_version` otherwise we will read and parse the version file twice.

eksperimental replied:
agreed!

josevalim commented on lib/elixir/lib/system.ex:
```diff
@@ -5,40 +5,71 @@ defmodule System do
+  # Tries to run "git rev-parse HEAD". In the case of success returns
+  # the revision hash. If that is not available, tries to read the commit hash
   # from .git/HEAD. If that fails, returns an empty string.
+  @spec get_revision :: String.t
```
It seems we never use the full version... so why bother not getting the short one?

eksperimental replied:
because we don't know if users want to user the full version for something else, like linking to the source file in github...
It'd better give the full version and letting them shorten it, than having them to do hacky stuck such as calling `git  rev-parse HEAD`.
but it's up to you the final word

josevalim replied:
@eksperimental we don't export the long get_revision anywhere else. This is not a public function. I don't see anywhere where the long version would be used.

eksperimental replied:
@josevalim what I ended up doing was having System.build_info[:revision] as the long version.
let me know if you are ok with that change

josevalim replied:
Isn't it backwards incompatible?

josevalim commented on lib/elixir/lib/system.ex:
```diff
@@ -73,11 +104,30 @@ defmodule System do
+      :version_build ->
+        info = build_info()
+        {:ok, v} = Version.parse(info[:version])
+        case v.pre do
```
This would also show the revision for non -dev stuff, like -rc.0, which we don't want. In general, I am :-1: on this feature. Folks can easily access it if they want by calling System.build_info.

eksperimental replied:
I created this function because we are using it it three times (elixir, iex, and mix), and we are duplicating code.
I just followed the [existing rationale](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/kernel/cli.ex#L155-L158). If you give me a detailed rationale I can fix this.

lexmag commented on lib/elixir/test/elixir/system_test.exs:
```diff
@@ -5,10 +5,26 @@ defmodule SystemTest do
+      assert String.length(build_info[:revision]) == 7
+    end
+
+    version_file = Path.join([__DIR__, "..", "..", "..", "..", "VERSION"]) |> Path.expand
```
I think we can write this as `Path.join(__DIR__, "../../../../VERSION"])`.

eksperimental replied:
but is that the right way to do it for Windows OS?

ericmj replied:
Yes, forward slash will work on all OSes.

lexmag commented on lib/elixir/lib/system.ex:
```diff
@@ -5,40 +5,71 @@ defmodule System do
+  defp strip(iodata) do
+    iodata
+    |> :re.replace("^[\s\r\n\t]+", "", [return: :binary])
+    |> :re.replace("[\s\r\n\t]+$", "", [return: :binary])
```
It'd be better to use single pattern here.

lexmag replied:
Maybe we should keep previous version of this function?
Previously it was more clear what we're replacing and in which case.

eksperimental replied:
previous version did not work for anything that was not a white spaces (that is. tabs, and carriage returns),
and neither it work when spaces were both at the beginning and the end of the file.
 
I cannot make it work squashing it into a single Regex.
```elixir
iex> "\nstrip this\n" |> :re.replace("^[\s\r\n\t]+|[\s\r\n\t]+$", "", [return: :binary])
"strip this\n"
```
but I'm sure someone with more knowledge of erlang can do it :-)

josevalim replied:
```elixir
"\nstrip this\n" |> :re.replace("^[\s\r\n\t]*(.*)[\s\r\n\t]*$", "\\1", [return: :binary])
```

lexmag replied:
Is there need to strip on line basis? I'm refereeing to `^` and `$` usage, which match start and end of the each line respectively (in contrast with `\A` and `\z` to match the whole string).

josevalim replied:
You are right, we could just remove all spaces in general.

eksperimental replied:
I wasn't able to make it work with \A \z, but this one does:
`"\t \nstrip this\n \r \n \t" |> :re.replace("[\s\r\n\t]*(.*)[\s\r\n\t]*", "\\1", [return: :binary])`

josevalim replied:
This is enough in our case: `:re.replace("[\s\r\n\t]", "", [:global, return: :binary])` because the version will never have spaces.

eksperimental replied:
the thing is that we are also using `strip` for `.git/HEAD`, and this will contain spaces.
actually, we need to fix that since .git/HEAD is a reference to the branch, and not the commit id itself.

eksperimental replied:
@lexmag and @josevalim: since strip was used in other files, and after different trials, I ended up with an improved version of the original function
```
:re.replace(iodata, "^[\s\r\n\t]+|[\s\r\n\t]+$", "", [:global, return: :binary])
```

lexmag commented on lib/elixir/lib/system.ex:
```diff
@@ -73,11 +104,30 @@ defmodule System do
+  @doc false
+  # Returns a string of the build info
+  @spec build_info(atom :: :version_build) :: String.t
+  def build_info(opt) when opt in [:version_build] do
```
We can match directly on `:version_build` here.

eksperimental replied:
you are 100% right

eksperimental commented:
thank you everyone for the comments, I think I have replied to every one of them. please let me know if you have more.

josevalim commented on lib/elixir/lib/system.ex:
```diff
@@ -73,11 +104,27 @@ defmodule System do
+
+  @doc false
+  # Returns a string of the build info
+  @spec build_info(atom :: :version_build) :: String.t
```
Let's get rid of this exported function please. We need to find a way to include this in the `build_info` otherwise we need to cut it out.

eksperimental replied:
thank @josevalim for the guidance.
I have moved everything in `System.build_info/0`
now that information can be access via `System.build_info[:version_build]`

eksperimental commented:
@josevalim: I have updated the PR, and fixed something new related to .git/HEAD
We need to define the rationale to when to show the short revision hash in the version (`System.build_info[:version_build]`),
and [the issue that requires git as a dependency](https://github.com/elixir-lang/elixir/pull/4296#discussion_r52904240)

josevalim commented on lib/elixir/lib/system.ex:
```diff
@@ -5,43 +5,64 @@ defmodule System do
+  @git_dir            Path.join(@base_dir, ".git")
+  @git_head_file      Path.join(@git_dir, "HEAD")
+
+  @spec strip(iodata) :: String.t
```
IIRC we don't spec private functions anywhere else in Elixir so I would drop those due to the inconsistency.

josevalim commented on lib/elixir/lib/system.ex:
```diff
@@ -5,43 +5,64 @@ defmodule System do
-          read_stripped(:filename.join(".git", "HEAD"))
+          case read_stripped(@git_head_file) do
+            "ref: " <> path ->
+              case read_stripped(Path.join(@git_dir, path)) do
```
Unecessary case.

josevalim commented:
I have added more comments. Let's please just drop reading the of the `.git/HEAD` file. It is adding more complexity without much benefits.

josevalim commented on lib/elixir/lib/system.ex:
```diff
@@ -73,11 +94,28 @@ defmodule System do
+    %{date:     get_date,
+      revision: revision,
+      version:  version,
+      version_build: version_build,
```
Let's call it just "build".

lexmag commented on lib/elixir/lib/system.ex:
```diff
@@ -5,43 +5,64 @@ defmodule System do
 
-  defp strip_re(iodata, pattern) do
-    :re.replace(iodata, pattern, "", [return: :binary])
+  @base_dir           Path.join([__DIR__, "../../.."])
```
We can use `Path.join/2` to get rid of the `[]`.

lexmag replied:
Also there is excessive indentation.

josevalim replied:
It would actually nice to not use `Path.join` function, the same way we didn't use it before, because using Path can put us into bootstrapping issues. Sure, it works now, but it may not work in a month if we decide to improve Path.

lexmag replied:
Very true.

lexmag commented on lib/elixir/lib/system.ex:
```diff
@@ -5,43 +5,64 @@ defmodule System do
+          :os.cmd('git rev-parse HEAD') |> strip
         else
-          read_stripped(:filename.join(".git", "HEAD"))
+          case read_stripped(@git_head_file) do
```
We can also replace this whole case with:
```elixir
with "ref: " <> path <- read_stripped(@git_head_file) do
  read_stripped(Path.join(@git_dir, path)) do
end
```

josevalim commented:
@eksperimental would be too much to ask to start over? This PR has a lot of changes and a lot of changes that should not really be added. We should avoid refactoring code when adding features. We should avoid adding multiple features at once. It makes it really hard to review and it makes really hard to see what we are gaining with each feature. For example, a lot of the code was written the way it is due to bootstrap reasons and, even if it works now, it may not work in the future. Those are changes that are better to avoid. The code also has backwards incompatible changes at this moment. Of course, the code also has good improvements but it has been really hard to separate them.

Here is my proposal:

1. Send a PR that fixes read_stripped
2. A PR that removes .git/HEAD support (it is wrong today and nobody found out, which means nobody used) and cleans up the makefile
3. A PR that adds a `:build` information to build_info and use it consistenly in Elixir and Mix

Is this reasonable?

josevalim commented:
Before I forget, thank you so much for those improvements. <3

eksperimental commented:
@josevalim yes, that's totally fine and it's even better like this. I will submit separate PRs in a few days time, and create issues for possible improvements.
@lexmag thank you for the comments, I will include them in the future PRs

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
